java/arcs/android/storage/database/DatabaseImpl.kt

```diff
@@ -98,7 +103,38 @@ class DatabaseImpl(
         data: DatabaseData,
         originatingClientId: Int?
     ): Int {
-        TODO("not implemented")
+        when (data) {
+            is DatabaseData.Entity -> insertOrUpdate(storageKey, data.entity)
+            else -> TODO("Support Singletons and Collections")
+        }
+        // TODO: Return a proper database version number.
+        return 1
+    }
+
+    @VisibleForTesting
+    suspend fun insertOrUpdate(storageKey: StorageKey, entity: Entity) {
+        writableDatabase.useTransaction {
```

Comment:
Nit: could make this a function expression, since all logic is done within the `useTransaction` block.

Reply:
Done